flink-connector-pinot/src/test/java/org/apache/flink/streaming/connectors/pinot/PinotSinkTest.java

```diff
@@ -0,0 +1,232 @@
+                        .setNumSlotsPerTaskManager(4)
+                        .setConfiguration(config)
+                        .build();
+
+        try (MiniCluster miniCluster = new MiniCluster(cfg)) {
+            miniCluster.start();
+            miniCluster.executeJobBlocking(jobGraph);
+        }
+    }
+
+    /**
+     * Sets up a DataStream using the provided execution environment and the provided input data.
+     *
+     * @param env  stream execution environment
+     * @param data Input data
+     */
+    private void setupDataStream(StreamExecutionEnvironment env, List<SingleColumnTableRow> data) {
+        // Create test stream
+        DataStream<SingleColumnTableRow> theData =
+                env.fromCollection(data)
+                        .name("Test input");
+
+        String tempDirPrefix = "flink-pinot-connector-test";
+        PinotSinkSegmentNameGenerator segmentNameGenerator = new SimpleSegmentNameGenerator(TABLE_NAME, "flink-connector");
+        FileSystemAdapter fsAdapter = new LocalFileSystemAdapter(tempDirPrefix);
+        JsonSerializer<SingleColumnTableRow> jsonSerializer = new SingleColumnTableRowSerializer();
+
+        EventTimeExtractor<SingleColumnTableRow> eventTimeExtractor = new SingleColumnTableRowEventTimeExtractor();
+
+        PinotSink<SingleColumnTableRow> sink = new PinotSink.Builder<SingleColumnTableRow>(getPinotHost(), getPinotControllerPort(), TABLE_NAME)
+                .withMaxRowsPerSegment(5)
+                .withTempDirectoryPrefix(tempDirPrefix)
+                .withJsonSerializer(jsonSerializer)
+                .withEventTimeExtractor(eventTimeExtractor)
+                .withSegmentNameGenerator(segmentNameGenerator)
+                .withFileSystemAdapter(fsAdapter)
+                .withNumCommitThreads(2)
+                .build();
+
+        // Sink into Pinot
+        theData.sinkTo(sink).name("Pinot sink");
+    }
+
+    /**
+     * As Pinot might take some time to index the recently pushed segments we might need to retry
+     * the {@link #checkForDataInPinot} method multiple times. This method provides a simple wrapper
+     * using linear retry backoff delay.
+     *
+     * @param data                  Data to expect in the Pinot table
+     * @param retryTimeoutInSeconds Maximum duration in seconds to wait for the data to arrive
+     * @throws InterruptedException
+     */
+    private void checkForDataInPinotWithRetry(List<SingleColumnTableRow> data, int retryTimeoutInSeconds) throws InterruptedException {
+        long endTime = System.currentTimeMillis() + 1000L * retryTimeoutInSeconds;
+        // Use max 10 retries with linear retry backoff delay
+        long retryDelay = 1000L / 10 * retryTimeoutInSeconds;
+        do {
+            try {
+                checkForDataInPinot(data);
+                // In case of no error, we can skip further retries
+                return;
+            } catch (AssertionFailedError | PinotControllerApiException e) {
```

Comment:
Do you still allow the `AssertionFailedError` I would assume once the job is finished all data is in pinot.

Reply:
It could still be that Pinot just hasn't (fully) finished indexing all the pushed segments when we check for the data.
Nonetheless, I spotted an error here: It could be that we catch the `AssertionFailedError`, sleep for some time and then finally do not execute the check once again as the `retryTimeoutInSeconds` was exceeded. So although the test actually should have failed (data not in Pinot), we would not raise an exception.

In order to prevent this, I've added a final check whether all data is in Pinot.